### DISCUSSION TITLE
morphTo relationship manager: list table columns specific for the morphed model?

### DISCUSSION THREAD

**Discussion by wivaku** [filamentphp/filament#7145] [Help]
I have a [one-to-many polymorphic relationship](https://laravel.com/docs/10.x/eloquent-relationships#one-to-many-polymorphic-relations): e.g. a Comment for either a Post or a Video.

For my `Comment` model I have method `$comment->commentable()` that returns either a `Post` or a `Video`.

For my Filament `CommentResource` I created a relationship manager `CommentableRelationManager`.

But how do define the table columns so that they show either the columns relevant for the `Post` or the ones relevant for the `Video`?

I could not find it in the [documentation](https://filamentphp.com/docs/2.x/admin/resources/relation-managers#listing-records).

**Updated**: my request may have been confusing. I want to define the table for my `CommentableRelationManager`. So, when I open a comment in my `CommentResource`, I see the related object that the comment was for: either a `Post` (with e.g. a textarea field) or a `Video` (with e.g. a dimensions field). How do I define the form/table in that `CommentableRelationManager`? Can it be used by both `Post` and `Video`?

**Comment by SimonMacIntyre** [comment]
I have the same question

**Comment by icanhazsoulz** [comment]
Looks line this is not documented, but the following does the trick (example from my code):
`TextColumn::make('commentable.full_name')
                    ->label('User')
                    ->searchable(),`

**Reply by micraux** [reply]
Seems you missed it in the [docs](https://filamentphp.com/docs/3.x/tables/getting-started#accessing-related-data-from-columns)

**Reply by joao-antonio-gomes** [reply]
But if video contains the attribute "views" and post contains the attribute "likes", how do you conditionally can show the attribute in same column?

**Comment by joao-antonio-gomes** [comment]
Hi buddy, I have the same problem and I found a workaround

You can create a method on morph model (Comment), who will deliver the attribute conditionally. You need to setup this method as attribute.

Source: https://laraveldaily.com/post/filament-table-dynamic-column-other-columns-accessor

Edit:

You can also do it this way:
```php
Tables\Columns\TextColumn::make('cpf_cnpj')
                    ->getStateUsing(fn($record) => $record->isPeople() ? $record->customerable->cpf : $record->customerable->cnpj)
```

**Reply by wivaku** [reply]
This is indeed a useful suggestion, and indeed a workaround.  
But hoping there's a way to do this not on the field level, but on the full table/form level. A la: 
```
// CommentableRelationManager

if(...) { // is there a way to determine the Model for that specific Commentable?
  return $table->[...Post...] 
} else {
  return $table->[...Video...] 
}
```

**Comment by BOFOII** [comment]
```
 Tables\Columns\TextColumn::make('licenable')->formatStateUsing(function ($record) {
                    return match ($record->licenable_type) {
                        SmmPanel::class => $record->licenable->name,
                        // todo
                    };
                }),
```

**Comment by georgknabl** [comment]
I stumbled upon this discussion because I had the same challenge. In the end I solved it like this:

My scenario: Several models use tags. I want a tag edit page to show all models this tag is connected to. In order for this to work, you need a TagTaggable pivot model.

My TaggableRelationManager attached to the TagResource:

```
class TaggablesRelationManager extends RelationManager
{
    protected static string $relationship = 'taggables';

    public function table(Table $table): Table
    {
        return $table
            ->recordTitleAttribute('taggable.name')
            ->columns([
                Tables\Columns\TextColumn::make('taggable.name')
                    ->label(lcl('Name'))
                    ->sortable()
                    ->searchable()
                    ->toggleable(),
                Tables\Columns\TextColumn::make('taggable_type')
                    ->label(lcl('Type'))
                    ->sortable()
                    ->searchable()
                    ->toggleable()
                    ->getStateUsing(fn ($record) => $record->taggable::getPluralModelLabel()),
            ])
            ->filters([
                Tables\Filters\SelectFilter::make('taggable_type')
                    ->label(lcl('Type'))
                    ->options(collect([
                        User::class,
                        UserGroup::class,
                        // ...
                    ])->mapWithKeys(fn ($class) => [
                        $class => $class::getPluralModelLabel(),
                    ])->toArray()),
            ])
            ->recordUrl(fn ($record) => route('filament.admin.resources.'.$record->taggable->getTable().'.edit', ['record' => $record->taggable->uuid]));
    }
}
```
